### PR TITLE
Implement a select box

### PR DESCRIPTION
**Pull request by DevYuns**
## Description

Implement  a basic `SelectBox` component.

#### Props
|                                 | necessary | types                            |  default                                     |
| ------------------ | :---------:    | -----------------------| :---------------------------------------------: |
| theme                      |                  | `DoobooTheme`                |                                      |
| data                         | ✓              | T[]                  |                                              |
| selectedIndex             | ✓                  | number                |         `0`                             |
| onSelect             | ✓                  |  (item: T, index: number) => void                |                                      |
| onPress             |                  |  () => void                |                                      |
| style                     |                   | `StyleProps<ViewStyle>`                |                                   |
| styles                      |                   | `Styles`             | `undefined`                                  |
| disabled         |           |          boolean                   |  `false`| 
| rightElement         |    |  ReactElement \| null      | `<Icon name="chevron-down-light"/>`                        |
| rotateDuration          |           | number                 | `200`                                        |
| isRightElemAnimated         |           |          boolean                   |  `true`|
|titleTouchableProps |        |   Partial<TouchableOpacityProps & GenericTouchableProps>  |     `{activeOpacity: 1} `          |
|itemTouchableProps |        |   Partial<TouchableOpacityProps & GenericTouchableProps>  |     `{activeOpacity: 1} `          |

 
ezgif com-gif-maker

### Issue

There is an issue that A `TouchableOpacity` in `absolute` view can not be clickable on `android`. I posted this issue [here](https://stackoverflow.com/questions/68916203/react-native-touchableopacity-in-absolute-view-can-not-be-clickable-on-andro).

#### update
When I set `pointerEvents` property to `box-none` to the parent view of `TouchableOpacity`, then touch event can reach to the right place.

But It is quite tricky, and can not works with `zIndex`. So it's hard to show dropdown over other component on `android`. but I'm getting closer to the solution.

#### resolved
I resolved this problem with `react-native-gesture-handler`. It works as expected on `android` as well as `ios` and `web`.

## Related Issues

N/A

## Tests

Render test & interaction test

## Checklist

Before you create this PR confirms that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide](https://github.com/dooboolab/dooboo-ui/blob/master/CONTRIBUTING.md) and followed the process outlined there for submitting PRs.
- [x] Run `yarn test` or `yarn test -u` if you need to update snapshot.
- [x] Run `yarn lint`
- [x] I am willing to follow-up on review comments in a timely manner.

### CODE REVIEW

**Comment by yujong-lee** [comment]
Great work! 

After first glance of code, I got these thoughts in mind.

1. I'm not sure, but have you think of `dropdown` as name? I want to hear your opinion on this!
2. I expect list should be collapsed on selection. I think at least option for this is needed.
3. Like #61, wrapping for mobile is needed.
4. Like #56, I think responsible for managing selected item and handling click should be removed.

How to you think?  😄

File: main/SelectBox/index.tsx [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+  itemHeight?: number;
+  translateYValue: number;
+  itemActiveOpacity: number;
+  onPress: (value: string) => void;
+}
+
+const ItemComp: FC<ItemCompProps> = ({
+  value,
+  styles,
+  itemHeight,
```
Comment (hyochan) [inline main/SelectBox/index.tsx:57]: Please refer to https://github.com/dooboolab/dooboo-ui/pull/67#discussion_r690609223. Never make style props as another props.

@yujong-lee I think it'd be awesome if we can have detailed code guidelines 😅 .

File: main/SelectBox/index.tsx [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+  data: string[];
+  theme?: DoobooTheme;
+  styles?: Styles;
+  rotateDuration?: number;
+  itemHeight?: number;
+  titleActiveOpacity?: number;
+  itemActiveOpacity?: number;
+  isRotate?: boolean;
+  hasRightElement?: boolean;
+  rightElement?: ReactElement;
```
Comment (hyochan) [inline main/SelectBox/index.tsx:94]: I think we only need `rightElement`

File: main/SelectBox/index.tsx [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+
+interface Props {
+  data: string[];
+  theme?: DoobooTheme;
+  styles?: Styles;
+  rotateDuration?: number;
+  itemHeight?: number;
+  titleActiveOpacity?: number;
+  itemActiveOpacity?: number;
+  isRotate?: boolean;
```
Comment (hyochan) [inline main/SelectBox/index.tsx:92]: shouldRotate?

File: main/SelectBox/index.tsx (hunk 1 above)
Comment (yujong-lee) [reply main/SelectBox/index.tsx:57]: @hyochan For sure. I will work on it little by little whenever I have time.

**Comment by DevYuns** [comment]
I modified some props according to reviews. and it works well on `ios` and `web`.  But there is a problem on `android`. When `TouchableOpacity` is in `absolute`, It can't get any press event. you can check related issue [here](https://github.com/facebook/react-native/issues/27232). I try to find a solution but it quite hard. Can someone help me?

File: main/SelectBox/index.tsx (hunk 1 above)
Comment (DevYuns) [reply main/SelectBox/index.tsx]: > Great work!
> 
> After first glance of code, I got these thoughts in mind.
> 
> 1. I'm not sure, but have you think of `dropdown` as name? I want to hear your opinion on this!
> 2. I expect list should be collapsed on selection. I think at least option for this is needed.
> 3. Like [fix: [CheckboxStory]  responsive layout #61](https://github.com/dooboolab/dooboo-ui/pull/61), wrapping for mobile is needed.
> 4. Like [Remove responsibility for managing state from [ButtonGroup] #56](https://github.com/dooboolab/dooboo-ui/pull/56), I think responsible for managing selected item and handling click should be removed.
> 
> How to you think? 😄
> 
> 스크린샷 2021-08-20 오후 10 46 35

I applied your suggestions. It's very good point. thanks.

**Comment by hyochan** [comment]
> I modified some props according to reviews. and it works well on `ios` and `web`. But there is a problem on `android`. When `TouchableOpacity` is in `absolute`, It can't get any press event. you can check related issue [here](https://github.com/facebook/react-native/issues/27232). I try to find a solution but it quite hard. Can someone help me?

Try to issue this in `stackoverflow` and share it in description.

**Comment by DevYuns** [comment]
I posted `android` issue on `stackoverflow`. You can see it [here](https://stackoverflow.com/questions/68916203/react-native-touchableopacity-in-absolute-view-can-not-be-clickable-on-andro).

**Review by hyochan — CHANGES_REQUESTED** [review]
Please add `peerDependency` at `lib/package.json`.
```
"react-native-gestuer-handler": "*"
```

**Comment by DevYuns** [comment]
@hyochan It's been a long time to test this component than I expected. I suffered from mocking `TouchableOpacity` from `react-native-gesture-handler`. anyway, I know this test code & test case are not the best. but.. I'll do better. Thank you for your patience.

**Review by hyochan — CHANGES_REQUESTED** [review]
I think you should `generic`. Supporting only `string` as a prop is very limited.

Imagine the below case.

```
<SelectBox
  data={data}
  style={{zIndex: 997}}
  onPress={(index) => {
    requestServer(item[index].id);
  }}
  selectedIndex={value3}
  rightElement={null}
/>
```

If `string` is the only data we can provide, how would user want to request API with data in `SelectBox`?

File: main/SelectBox/index.tsx [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+export interface Props {
+  data: string[];
+  onPress?: (i: number) => void;
+  selectedIndex?: number;
+  style?: StyleProp<ViewStyle>;
+  styles?: Styles;
+  rotateDuration?: number;
+  titleActiveOpacity?: number;
+  itemActiveOpacity?: number;
+  shouldRotate?: boolean;
```
Comment (hyochan) [inline main/SelectBox/index.tsx:95]: We can't think of this as an `animated` props at a glance. Suggest you to rename this

File: main/SelectBox/index.tsx [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+  onPress?: (i: number) => void;
+}
+
+const ItemComp: FC<ItemCompProps & {theme: DoobooTheme}> = ({
+  value,
+  order,
+  styles,
+  theme,
+  setIsOpened,
+  itemActiveOpacity,
```
Comment (hyochan) [inline main/SelectBox/index.tsx:62]: How about supporting a bigger concept here.
`itemTouchableProps`?

File: main/SelectBox/index.tsx [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+};
+
+export interface Props {
+  data: string[];
+  onPress?: (i: number) => void;
+  selectedIndex?: number;
+  style?: StyleProp<ViewStyle>;
+  styles?: Styles;
+  rotateDuration?: number;
+  titleActiveOpacity?: number;
```
Comment (hyochan) [inline main/SelectBox/index.tsx:93]: Same here. `titleTouchableProps`

**Review by hyochan — CHANGES_REQUESTED** [review]
Also please add `readme` for the new component!

**Comment by DevYuns** [comment]
@hyochan 현재 리팩토링은 다했고 `readme`도 작성했는데, 마지막에 컴포넌트를 `mdx` 파일에 올리니까 아래와 같은 에러가 자꾸 발생해서 알아보고 있습니다.

```
Uncaught Error: Invalid hook call. Hooks can only be called inside of the body of a function component. This could happen for one of the following reasons:
1. You might have mismatching versions of React and the renderer (such as React DOM)
2. You might be breaking the Rules of Hooks
3. You might have more than one copy of React in the same app
See https://reactjs.org/link/invalid-hook-call for tips about how to debug and fix this problem.
```

**Review by hyochan — APPROVED** [review]
Great work~! 💯

**Comment by hyochan** [comment]
> @hyochan I completed a refactoring and wrote `readme`, but I got an error like below when adding a component to the `mdx` file to show a Demo version. I think that It's because `react-native-gesture-handler` but not sure.
> 
> ```
> Uncaught Error: Invalid hook call. Hooks can only be called inside of the body of a function component. This could happen for one of the following reasons:
> 1. You might have mismatching versions of React and the renderer (such as React DOM)
> 2. You might be breaking the Rules of Hooks
> 3. You might have more than one copy of React in the same app
> See https://reactjs.org/link/invalid-hook-call for tips about how to debug and fix this problem.
> ```

You are right~! you need to add `react-native-gesture-handler`. I just did it for you.

**Comment by hyochan** [comment]
@DevYuns Next time please write good commit messages or good descriptions in `PR` content.
There weren't meaningful messages when merging current `PR`.

**Pull request merged**